### PR TITLE
Add Scalar::to_(u|i)16 methods

### PR DESCRIPTION
**Pull request by pvdrz**
r? @RalfJung

### CODE REVIEW

File: src/librustc/mir/interpret/value.rs [hunk 1]
```diff
@@ -452,6 +458,13 @@ impl<'tcx, Tag> Scalar<Tag> {
         Ok(b as i32)
     }
 
+    pub fn to_i16(self) -> InterpResult<'static, i16> {
```
Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]: Please keep them sorted: `i16` should be between `i8` and `i32`

Comment (christianpoveda) [reply src/librustc/mir/interpret/value.rs]: My bad, I've corrected it

File: src/librustc/mir/interpret/value.rs [hunk 2]
```diff
@@ -421,6 +421,12 @@ impl<'tcx, Tag> Scalar<Tag> {
         Ok(b as u8)
     }
 
+    pub fn to_u16(self) -> InterpResult<'static, u16> {
```
Comment (Centril) [inline src/librustc/mir/interpret/value.rs]: While we're at it... could you sprinkle doc comments on the methods in this PR (or another...)?

File: src/librustc/mir/interpret/value.rs [hunk 3]
```diff
@@ -445,6 +451,13 @@ impl<'tcx, Tag> Scalar<Tag> {
         Ok(b as i8)
     }
 
+    pub fn to_i16(self) -> InterpResult<'static, i16> {
+        let sz = Size::from_bits(16);
+        let b = self.to_bits(sz)?;
+        let b = sign_extend(b, sz) as i128;
+        Ok(b as i16)
+    }
+
```
Comment (Centril) [inline src/librustc/mir/interpret/value.rs]: Why not have `to_signed_with_bit_width` (?) and then rewrite this as:
```rust
self.to_signed_with_bit_width(16).map(|v| v as i16)
```
(Same with the other types and unsigned variants.)

Comment (christianpoveda) [reply src/librustc/mir/interpret/value.rs]: done :D

File: src/librustc/mir/interpret/value.rs (hunk 2 above)
Comment (christianpoveda) [reply src/librustc/mir/interpret/value.rs]: How about 
```    
// Consume the scalar to produce an `(u|i)XX`, it fails if the scalar is a pointer .
```

Comment (Centril) [reply src/librustc/mir/interpret/value.rs]: New sentence at "It fails", otherwise sgtm.

File: src/librustc/mir/interpret/value.rs [hunk 4]
```diff
@@ -415,48 +415,60 @@ impl<'tcx, Tag> Scalar<Tag> {
         }
     }
 
+    fn to_unsigned_with_bit_width(self, bits: u64) -> InterpResult<'static, u128> {
+        let sz = Size::from_bits(bits);
+        self.to_bits(sz)
+    }
+
+    // Consume the scalar to produce an `u8`. It fails if the scalar is a pointer.
```
Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]: "Consume" isn't really relevant here, so it shouldn't be what the docs lead with. Also the grammar is inconsistent between the two sentences.

What about: "Converts the scalar to produce a `u8`. Fails if the scalar is a pointer."

File: src/librustc/mir/interpret/value.rs [hunk 5]
```diff
@@ -415,48 +415,60 @@ impl<'tcx, Tag> Scalar<Tag> {
         }
     }
 
+    fn to_unsigned_with_bit_width(self, bits: u64) -> InterpResult<'static, u128> {
```
Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]: Please mark this as `#[inline]`

File: src/librustc/mir/interpret/value.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: Also also make sure these are all actually doc comments: They should start with `///`

Comment (christianpoveda) [reply src/librustc/mir/interpret/value.rs]: Done :P

File: src/librustc/mir/interpret/value.rs (hunk 5 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: @christianpoveda you saw this comment? (just wondering as you reacted to the other one and then stopped, no rush)

Comment (christianpoveda) [reply src/librustc/mir/interpret/value.rs]: It seems I missed it :P

**Comment by RalfJung** [comment]
LGTM, thanks!

@bors r+

**Comment by RalfJung** [comment]
@christianpoveda can you submit a PR against Miri to use this function and remove the FIXME?

**Comment by christianpoveda** [comment]
sure

**Pull request merged**